c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -31,40 +31,79 @@ use rustc_middle::ty::{SubstsRef, Ty, TyKind};
     /// Subset constraints on pointer permissions.  For example, this contains constraints like
     /// "the `PermissionSet` assigned to `PointerId` `l1` must be a subset of the `PermissionSet`
     /// assigned to `l2`".  See `dataflow::Constraint` for a full description of supported
     /// constraints.
-    constraints: DataflowConstraints,
+    constraints: Option<DataflowConstraints>,
     /// Equivalence constraints on pointer permissions and flags.  An entry `(l1, l2)` in this list
     /// means that `PointerId`s `l1` and `l2` should be assigned exactly the same permissions and
     /// flags.  This ensures that the two pointers will be rewritten to the same safe type.
     ///
     /// Higher-level code eventually feeds the constraints recorded here into the union-find data
     /// structure defined in `crate::equiv`, so adding a constraint here has the effect of unifying
     /// the equivalence classes of the two `PointerId`s.
-    equiv_constraints: Vec<(PointerId, PointerId)>,
+    equiv_constraints: Option<Vec<(PointerId, PointerId)>>,
 }
 
 impl<'tcx> TypeChecker<'tcx, '_> {
     fn add_edge(&mut self, src: PointerId, dest: PointerId) {
         // Copying `src` to `dest` can discard permissions, but can't add new ones.
-        self.constraints.add_subset(dest, src);
+        if let Some(ref mut constraints) = self.constraints {
+            constraints.add_subset(dest, src);
+        }
     }
 
     fn add_edge_except(&mut self, src: PointerId, dest: PointerId, except: PermissionSet) {
         // Copying `src` to `dest` can discard permissions, but can't add new ones,
         // except for the specified exceptions.
-        self.constraints.add_subset_except(dest, src, except);
+        if let Some(ref mut constraints) = self.constraints {
+            constraints.add_subset_except(dest, src, except);
+        }
+    }
+
+    /// Add `Constraint::AllPerms`, which requires `ptr` to have all of the permissions listed in
+    /// `perms`.
+    fn add_all_perms(&mut self, ptr: PointerId, perms: PermissionSet) {
+        if let Some(ref mut constraints) = self.constraints {
+            constraints.add_all_perms(ptr, perms);
+        }
+    }
+
+    /// Add `Constraint::NoPerms`, which requires `ptr` to have none of the permissions listed in
+    /// `perms`.
+    fn add_no_perms(&mut self, ptr: PointerId, perms: PermissionSet) {
+        if let Some(ref mut constraints) = self.constraints {
+            constraints.add_no_perms(ptr, perms);
+        }
     }
 
     fn add_equiv(&mut self, a: PointerId, b: PointerId) {
-        self.equiv_constraints.push((a, b));
+        if let Some(ref mut equiv_constraints) = self.equiv_constraints {
+            equiv_constraints.push((a, b));
+        }
+    }
+
+    fn pointee_type(&self, ptr: PointerId) -> Option<LTy<'tcx>> {
+        self.pointee_types
+            .as_ref()
+            .and_then(|pointee_types| pointee_types[ptr].get_sole_lty())
```

Comment:
Will this panic if `ptr` is out of bounds?

Reply:
It will panic if `ptr` is `PointerId::NONE` (the same is true of the code this is replacing).  I added an assert just to make it clear that this should only be called on actual pointers: 7f1d501